// nitro-nightly@4.0.0-2025101/dist/cli/dev.mjs
import { d as defineCommand } from './index.mjs';
import { consola } from 'consola';
import { createNitro, prepare, build } from 'nitro';
import { c as commonArgs } from './common.mjs';
import { N as NitroDevServer } from '../_chunks/server.mjs';
import { r as resolve } from '../_chunks/pathe.M-eThtNZ.mjs';
import 'consola/utils';
import 'nitro/meta';
import '../_chunks/app.mjs';
import 'node:fs';
import 'node:fs/promises';
import 'node:worker_threads';
import 'std-env';
import 'fs';
import 'fs/promises';
import 'events';
import 'path';
import 'node:stream';
import 'node:path';
import 'os';
import 'h3';
import 'url';
import 'tty';
import 'util';
import 'crypto';
import 'stream';
import 'ufo';
import 'node:http';
import 'node:https';
import 'node:events';
import 'undici';
import 'youch-core';
import 'youch';
import 'source-map';
import 'srvx';
import 'srvx/node';

const hmrKeyRe = /^runtimeConfig\.|routeRules\./;
const dev = defineCommand({
  meta: {
    name: "dev",
    description: "Start the development server"
  },
  args: {
    ...commonArgs,
    port: { type: "string", description: "specify port" },
    host: { type: "string", description: "specify hostname " }
  },
  async run({ args }) {
    const rootDir = resolve(args.dir || args._dir || ".");
    let nitro;
    const reload = async () => {
      if (nitro) {
        consola.info("Restarting dev server...");
        if ("unwatch" in nitro.options._c12) {
          await nitro.options._c12.unwatch();
        }
        await nitro.close();
      }
      nitro = await createNitro(
        {
          rootDir,
          dev: true,
          _cli: { command: "dev" }
        },
        {
          watch: true,
          c12: {
            async onUpdate({ getDiff, newConfig }) {
              const diff = getDiff();
              if (diff.length === 0) {
                return;
              }
              consola.info(
                "Nitro config updated:\n" + diff.map((entry) => `  ${entry.toString()}`).join("\n")
              );
              await (diff.every((e) => hmrKeyRe.test(e.key)) ? nitro.updateConfig(newConfig.config || {}) : reload());
            }
          }
        }
      );
      nitro.hooks.hookOnce("restart", reload);
      const server = new NitroDevServer(nitro);
      await server.listen({
        port: args.port,
        hostname: args.host
      });
      await prepare(nitro);
      await build(nitro);
    };
    await reload();
  }
});

export { dev as default };
