// nitro-nightly@4.0.0-2025101/dist/_chunks/server.mjs
import { N as NitroDevApp, d as debounce, w as watch, a as NodeDevWorker } from './app.mjs';
import { HTTPError } from 'h3';
import { version } from 'nitro/meta';
import consola from 'consola';
import { writeFile } from 'node:fs/promises';
import { serve } from 'srvx/node';
import { isTest, isCI } from 'std-env';
import { r as resolve } from './pathe.M-eThtNZ.mjs';

function createDevServer(nitro) {
  return new NitroDevServer(nitro);
}
class NitroDevServer extends NitroDevApp {
  #entry;
  #workerData = {};
  #listeners = [];
  #watcher;
  #workers = [];
  #workerIdCtr = 0;
  #workerError;
  #building = true;
  // Assume initial build will start soon
  #buildError;
  #messageListeners = /* @__PURE__ */ new Set();
  constructor(nitro) {
    super(nitro, async (event) => {
      const worker = await this.#getWorker();
      if (!worker) {
        return this.#generateError();
      }
      return worker.fetch(event.req);
    });
    for (const key of Object.getOwnPropertyNames(NitroDevServer.prototype)) {
      const value = this[key];
      if (typeof value === "function" && key !== "constructor") {
        this[key] = value.bind(this);
      }
    }
    this.#entry = resolve(
      nitro.options.output.dir,
      nitro.options.output.serverDir,
      "index.mjs"
    );
    nitro.hooks.hook("close", () => this.close());
    nitro.hooks.hook("dev:start", () => {
      this.#building = true;
      this.#buildError = void 0;
    });
    nitro.hooks.hook("dev:reload", (payload) => {
      this.#buildError = void 0;
      this.#building = false;
      if (payload?.entry) {
        this.#entry = payload.entry;
      }
      if (payload?.workerData) {
        this.#workerData = payload.workerData;
      }
      this.reload();
    });
    nitro.hooks.hook("dev:error", (cause) => {
      this.#buildError = cause;
      this.#building = false;
      for (const worker of this.#workers) {
        worker.close();
      }
    });
    if (nitro.options.devServer.watch.length > 0) {
      const debouncedReload = debounce(() => this.reload());
      this.#watcher = watch(
        nitro.options.devServer.watch,
        nitro.options.watchOptions
      );
      this.#watcher.on("add", debouncedReload).on("change", debouncedReload);
    }
  }
  // #region Public Methods
  async upgrade(req, socket, head) {
    const worker = await this.#getWorker();
    if (!worker) {
      throw new HTTPError({
        status: 503,
        statusText: "No worker available."
      });
    }
    return worker.upgrade(req, socket, head);
  }
  listen(opts) {
    const server = serve({
      ...opts,
      fetch: this.fetch
    });
    this.#listeners.push(server);
    if (server.node?.server) {
      server.node.server.on(
        "upgrade",
        (req, sock, head) => this.upgrade(req, sock, head)
      );
    }
    return server;
  }
  async close() {
    await Promise.all(
      [
        Promise.all(this.#listeners.map((l) => l.close())).then(() => {
          this.#listeners = [];
        }),
        Promise.all(this.#workers.map((w) => w.close())).then(() => {
          this.#workers = [];
        }),
        Promise.resolve(this.#watcher?.close()).then(() => {
          this.#watcher = void 0;
        })
      ].map(
        (p) => p.catch((error) => {
          consola.error(error);
        })
      )
    );
  }
  reload() {
    for (const worker2 of this.#workers) {
      worker2.close();
    }
    const worker = new NodeDevWorker({
      name: `Nitro_${this.#workerIdCtr++}`,
      entry: this.#entry,
      data: {
        ...this.#workerData,
        globals: {
          __NITRO_RUNTIME_CONFIG__: this.nitro.options.runtimeConfig,
          ...this.#workerData.globals
        }
      },
      hooks: {
        onClose: (worker2, cause) => {
          this.#workerError = cause;
          const index = this.#workers.indexOf(worker2);
          if (index !== -1) {
            this.#workers.splice(index, 1);
          }
        },
        onReady: (worker2, addr) => {
          this.#writeBuildInfo(worker2, addr);
        }
      }
    });
    if (!worker.closed) {
      for (const listener of this.#messageListeners) {
        worker.onMessage(listener);
      }
      this.#workers.unshift(worker);
    }
  }
  sendMessage(message) {
    for (const worker of this.#workers) {
      if (!worker.closed) {
        worker.sendMessage(message);
      }
    }
  }
  onMessage(listener) {
    this.#messageListeners.add(listener);
    for (const worker of this.#workers) {
      worker.onMessage(listener);
    }
  }
  offMessage(listener) {
    this.#messageListeners.delete(listener);
    for (const worker of this.#workers) {
      worker.offMessage(listener);
    }
  }
  // #endregion
  // #region Private Methods
  #writeBuildInfo(_worker, addr) {
    const buildInfoPath = resolve(this.nitro.options.buildDir, "nitro.json");
    const buildInfo = {
      date: (/* @__PURE__ */ new Date()).toJSON(),
      preset: this.nitro.options.preset,
      framework: this.nitro.options.framework,
      versions: {
        nitro: version
      },
      dev: {
        pid: process.pid,
        workerAddress: addr
      }
    };
    writeFile(buildInfoPath, JSON.stringify(buildInfo, null, 2)).catch(
      (error) => {
        consola.error(error);
      }
    );
  }
  async #getWorker() {
    let retry = 0;
    const maxRetries = isTest || isCI ? 100 : 10;
    while (this.#building || ++retry < maxRetries) {
      if ((this.#workers.length === 0 || this.#buildError) && !this.#building) {
        return;
      }
      const activeWorker = this.#workers.find((w) => w.ready);
      if (activeWorker) {
        return activeWorker;
      }
      await new Promise((resolve2) => setTimeout(resolve2, 600));
    }
  }
  #generateError() {
    const error = this.#buildError || this.#workerError;
    if (error) {
      try {
        error.unhandled = false;
        let id = error.id || error.path;
        if (id) {
          const cause = error.errors?.[0];
          const loc = error.location || error.loc || cause?.location || cause?.loc;
          if (loc) {
            id += `:${loc.line}:${loc.column}`;
          }
          error.stack = (error.stack || "").replace(
            /(^\s*at\s+.+)/m,
            `    at ${id}
$1`
          );
        }
      } catch {
      }
      return new HTTPError(error);
    }
    return new Response(
      JSON.stringify(
        {
          error: "Dev server is unavailable.",
          hint: "Please reload the page and check the console for errors if the issue persists."
        },
        null,
        2
      ),
      {
        status: 503,
        statusText: "Dev server is unavailable",
        headers: {
          "Content-Type": "application/json",
          "Cache-Control": "no-store",
          Refresh: "3"
        }
      }
    );
  }
  // #endregion
}

export { NitroDevServer as N, createDevServer as c };
