// nitro-nightly@4.0.0-2025101/dist/runtime/internal/route-rules.mjs
import { proxyRequest, redirect as sendRedirect } from "h3";
import { joinURL, withQuery, withoutBase } from "ufo";
import { defineCachedEventHandler } from "./cache.mjs";
export const headers = ((m) => function headersRouteRule(event) {
  for (const [key, value] of Object.entries(m.options || {})) {
    event.res.headers.set(key, value);
  }
});
export const redirect = ((m) => function redirectRouteRule(event) {
  let target = m.options?.to;
  if (!target) {
    return;
  }
  if (target.endsWith("/**")) {
    let targetPath = event.url.pathname + event.url.search;
    const strpBase = m.options._redirectStripBase;
    if (strpBase) {
      targetPath = withoutBase(targetPath, strpBase);
    }
    target = joinURL(target.slice(0, -3), targetPath);
  } else if (event.url.search) {
    target = withQuery(target, Object.fromEntries(event.url.searchParams));
  }
  return sendRedirect(target, m.options?.status);
});
export const proxy = ((m) => function proxyRouteRule(event) {
  let target = m.options?.to;
  if (!target) {
    return;
  }
  if (target.endsWith("/**")) {
    let targetPath = event.url.pathname + event.url.search;
    const strpBase = m.options._proxyStripBase;
    if (strpBase) {
      targetPath = withoutBase(targetPath, strpBase);
    }
    target = joinURL(target.slice(0, -3), targetPath);
  } else if (event.url.search) {
    target = withQuery(target, Object.fromEntries(event.url.searchParams));
  }
  return proxyRequest(event, target, {
    ...m.options
  });
});
export const cache = ((m) => function cacheRouteRule(event, next) {
  if (!event.context.matchedRoute) {
    return next();
  }
  const cachedHandlers = globalThis.__nitroCachedHandlers ??= /* @__PURE__ */ new Map();
  const { handler, route } = event.context.matchedRoute;
  const key = `${m.route}:${route}`;
  let cachedHandler = cachedHandlers.get(key);
  if (!cachedHandler) {
    cachedHandler = defineCachedEventHandler(handler, {
      group: "nitro/route-rules",
      name: key,
      ...m.options
    });
    cachedHandlers.set(key, cachedHandler);
  }
  return cachedHandler(event);
});
