// @tanstack/router-ssr-query-core/dist/cjs/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
let _tanstack_query_core = require("@tanstack/query-core");
let _tanstack_router_core = require("@tanstack/router-core");
let _tanstack_router_core_isServer = require("@tanstack/router-core/isServer");
//#region src/index.ts
function setupCoreRouterSsrQueryIntegration({ router, queryClient, dehydrateOptions, hydrateOptions, handleRedirects = true }) {
	const ogHydrate = router.options.hydrate;
	const ogDehydrate = router.options.dehydrate;
	if (_tanstack_router_core_isServer.isServer ?? router.isServer) {
		const sentQueries = /* @__PURE__ */ new Set();
		const queryStream = createPushableStream();
		let unsubscribe = void 0;
		let cleanupRegistered = false;
		let tornDown = false;
		const teardown = () => {
			if (tornDown) return;
			tornDown = true;
			try {
				unsubscribe?.();
			} catch {}
			unsubscribe = void 0;
			try {
				if (!queryStream.isClosed()) queryStream.close();
			} catch {}
			try {
				queryClient.cancelQueries();
			} catch {}
			try {
				queryClient.clear();
			} catch {}
			sentQueries.clear();
		};
		const registerCleanup = (serverSsr = router.serverSsr) => {
			if (cleanupRegistered) return;
			if (!serverSsr) return;
			serverSsr.onCleanup(teardown);
			cleanupRegistered = true;
		};
		router.serverSsrLifecycle = {
			...router.serverSsrLifecycle,
			onServerSsrAttach: [...router.serverSsrLifecycle?.onServerSsrAttach ?? [], registerCleanup]
		};
		router.options.dehydrate = async () => {
			router.serverSsr.onRenderFinished(() => {
				if (!queryStream.isClosed()) queryStream.close();
				unsubscribe?.();
				unsubscribe = void 0;
			});
			const dehydratedRouter = {
				...await ogDehydrate?.(),
				queryStream: queryStream.stream
			};
			const dehydratedQueryClient = (0, _tanstack_query_core.dehydrate)(queryClient, dehydrateOptions);
			if (dehydratedQueryClient.queries.length > 0) {
				dehydratedQueryClient.queries.forEach((query) => {
					sentQueries.add(query.queryHash);
				});
				dehydratedRouter.dehydratedQueryClient = dehydratedQueryClient;
			}
			return dehydratedRouter;
		};
		const ogClientOptions = queryClient.getDefaultOptions();
		queryClient.setDefaultOptions({
			...ogClientOptions,
			dehydrate: {
				shouldDehydrateQuery: () => true,
				...ogClientOptions.dehydrate
			}
		});
		unsubscribe = queryClient.getQueryCache().subscribe((event) => {
			if (!router.serverSsr?.isDehydrated()) return;
			if (sentQueries.has(event.query.queryHash)) return;
			if (!event.query.promise) return;
			if (queryStream.isClosed()) {
				console.warn(`tried to stream query ${event.query.queryHash} after stream was already closed`);
				return;
			}
			const dehydratedQuery = (0, _tanstack_query_core.dehydrate)(queryClient, {
				...dehydrateOptions,
				shouldDehydrateQuery: (query) => {
					if (query.queryHash !== event.query.queryHash) return false;
					return (ogClientOptions.dehydrate?.shouldDehydrateQuery?.(query) ?? true) && (dehydrateOptions?.shouldDehydrateQuery?.(query) ?? true);
				}
			});
			if (dehydratedQuery.queries.length === 0) return;
			sentQueries.add(event.query.queryHash);
			queryStream.enqueue(dehydratedQuery);
		});
	} else {
		router.options.hydrate = async (dehydrated) => {
			await ogHydrate?.(dehydrated);
			if (dehydrated.dehydratedQueryClient) (0, _tanstack_query_core.hydrate)(queryClient, dehydrated.dehydratedQueryClient, hydrateOptions);
			const reader = dehydrated.queryStream.getReader();
			reader.read().then(async function handle({ done, value }) {
				(0, _tanstack_query_core.hydrate)(queryClient, value, hydrateOptions);
				if (done) return;
				return handle(await reader.read());
			}).catch((err) => {
				console.error("Error reading query stream:", err);
			});
		};
		if (handleRedirects) {
			const ogMutationCacheConfig = queryClient.getMutationCache().config;
			queryClient.getMutationCache().config = {
				...ogMutationCacheConfig,
				onError: (error, ...rest) => {
					if ((0, _tanstack_router_core.isRedirect)(error)) {
						error.options._fromLocation = router.stores.location.get();
						return router.navigate(router.resolveRedirect(error).options);
					}
					return ogMutationCacheConfig.onError?.(error, ...rest);
				}
			};
			const ogQueryCacheConfig = queryClient.getQueryCache().config;
			queryClient.getQueryCache().config = {
				...ogQueryCacheConfig,
				onError: (error, ...rest) => {
					if ((0, _tanstack_router_core.isRedirect)(error)) {
						error.options._fromLocation = router.stores.location.get();
						return router.navigate(router.resolveRedirect(error).options);
					}
					return ogQueryCacheConfig.onError?.(error, ...rest);
				}
			};
		}
	}
}
function createPushableStream() {
	let controllerRef;
	const stream = new ReadableStream({ start(controller) {
		controllerRef = controller;
	} });
	let _isClosed = false;
	return {
		stream,
		enqueue: (chunk) => {
			if (!_isClosed) controllerRef.enqueue(chunk);
		},
		close: () => {
			if (_isClosed) return;
			controllerRef.close();
			_isClosed = true;
		},
		isClosed: () => _isClosed,
		error: (err) => {
			if (_isClosed) return;
			_isClosed = true;
			controllerRef.error(err);
		}
	};
}
//#endregion
exports.setupCoreRouterSsrQueryIntegration = setupCoreRouterSsrQueryIntegration;
