// nitro-nightly@4.0.0-2025101/dist/presets/azure/runtime/azure-swa.mjs
import "#nitro-internal-pollyfills";
import { parseURL } from "ufo";
import { useNitroApp } from "nitro/runtime";
import { getAzureParsedCookiesFromHeaders } from "./_utils.mjs";
const nitroApp = useNitroApp();
export async function handle(context, req) {
  let url;
  if (req.headers["x-ms-original-url"]) {
    const parsedURL = parseURL(req.headers["x-ms-original-url"]);
    url = parsedURL.pathname + parsedURL.search;
  } else {
    url = "/api/" + (req.params.url || "");
  }
  const response = await nitroApp.fetch(url, {
    method: req.method || void 0,
    // https://github.com/Azure/azure-functions-nodejs-worker/issues/294
    // https://github.com/Azure/azure-functions-host/issues/293
    body: req.bufferBody ?? req.rawBody
  });
  context.res = {
    status: response.status,
    body: response.body,
    cookies: getAzureParsedCookiesFromHeaders(response.headers),
    headers: Object.fromEntries(
      [...response.headers.entries()].filter(([key]) => key !== "set-cookie")
    )
  };
}
