// nitro-nightly@4.0.0-2025101/dist/presets/cloudflare/runtime/_module-handler.mjs
import "#nitro-internal-pollyfills";
import { useNitroApp } from "nitro/runtime";
import { runCronTasks } from "nitro/runtime/internal";
export function createHandler(hooks) {
  const nitroApp = useNitroApp();
  return {
    async fetch(request, env, context) {
      const ctxExt = {};
      const url = new URL(request.url);
      if (hooks.fetch) {
        const res = await hooks.fetch(request, env, context, url, ctxExt);
        if (res) {
          return res;
        }
      }
      return fetchHandler(request, env, context, url, nitroApp, ctxExt);
    },
    scheduled(controller, env, context) {
      globalThis.__env__ = env;
      context.waitUntil(
        nitroApp.hooks.callHook("cloudflare:scheduled", {
          controller,
          env,
          context
        })
      );
      if (import.meta._tasks) {
        context.waitUntil(
          runCronTasks(controller.cron, {
            context: {
              cloudflare: {
                env,
                context
              }
            },
            payload: {}
          })
        );
      }
    },
    email(message, env, context) {
      globalThis.__env__ = env;
      context.waitUntil(
        nitroApp.hooks.callHook("cloudflare:email", {
          message,
          event: message,
          // backward compat
          env,
          context
        })
      );
    },
    queue(batch, env, context) {
      globalThis.__env__ = env;
      context.waitUntil(
        nitroApp.hooks.callHook("cloudflare:queue", {
          batch,
          event: batch,
          env,
          context
        })
      );
    },
    tail(traces, env, context) {
      globalThis.__env__ = env;
      context.waitUntil(
        nitroApp.hooks.callHook("cloudflare:tail", {
          traces,
          env,
          context
        })
      );
    },
    trace(traces, env, context) {
      globalThis.__env__ = env;
      context.waitUntil(
        nitroApp.hooks.callHook("cloudflare:trace", {
          traces,
          env,
          context
        })
      );
    }
  };
}
export async function fetchHandler(cfReq, env, context, url = new URL(cfReq.url), nitroApp = useNitroApp(), ctxExt) {
  globalThis.__env__ = env;
  const req = cfReq;
  req.runtime ??= { name: "cloudflare" };
  req.runtime.cloudflare ??= { context, env };
  req.waitUntil = context.waitUntil.bind(context);
  return nitroApp.fetch(req);
}
